pandas/core/internals/__init__.py

```diff
@@ -2785,6 +2787,17 @@ def set(self, locs, values, check=False):
 
         self.values[locs] = values
 
+    def eval(self, func, other, try_cast=False, **kwargs):
+        block = super(DatetimeBlock, self).eval(func, other, try_cast=try_cast,
```

Comment:
What happens here if instead of calling super eval you delegate to self.holder?

Reply:
In `DatetimeBlock - datetime` case and `DatetimeBlock + DateOffset` case, it require the result block changing (`DatetimeBlock` -> `TimeDeltaBlock `, `ObjectBlock` -> `DatetimeBlock` or `DatetimeTZBlock`). 
Delegating to `self.holder` couldn't change block type. Besides, `self._holder` is unaware of `other`.

In fact, what should be blamed is the last line in `Block.eval` carelessly make a wrong block. A proper fix should be 
1. spliting a method to make block regarding to `other`.
2. override `make_block` with `other` as argument. But this will lead to `self.make_block` returning another type of block.
https://github.com/pandas-dev/pandas/blob/0828c25913b5862d70322471a6ee4fa11d8f890d/pandas/core/internals/__init__.py#L1456

Reply:
> In fact, what should be blamed is the last line in Block.eval carelessly make a wrong block. A proper fix should be

There has been an effort recently to fix mismatches between the behaviors of arithmetic ops in Index vs Series vs DataFrame.  For the timedelta and datetime cases, the canonical behavior is defined in TimedeltaIndex/DatetimeIndex.

The _best_ case here to to avoid making this fix in `core.internals` but instead do it in core.DataFrame (or core.ops).  The next base case is to have the FooBlock dispatch to the appropriate Index subclass to ensure the behavior is canonical.

Reply:
That will be great, but it may need a big rewrite of `eval` and `_try_coerce_args`, cause now `values` is all coerced to `ndarray[i8]` for `DatetimeLike`. (May be benefited from numpy bi-ops, but confused to convert back.)

Is there an issue to xfer? This PR would depand on it.